File: Leet 2657/solution.py
class Solution:
    def findThePrefixCommonArray(self, A: list[int], B: list[int]) -> list[int]:
        length: int = min(len(A), len(B))
        tally: dict[int, int] = {}
        result: list[int] = []

        count: int = 0

        for i in range(length):
            if A[i] not in tally:
                tally[A[i]] = 0

            tally[A[i]] += 1
            if tally[A[i]] == 2:
                count += 1
            
            if B[i] not in tally:
                tally[B[i]] = 0

            tally[B[i]] += 1
            if tally[B[i]] == 2:
                count += 1
                
            result.append(count)

        return result


def main() -> None:
    print("2657. Find the Prefix Common Array of Two Arrays")

    sol = Solution()
    print(sol.findThePrefixCommonArray([1,3,2,4], [3,1,2,4]))
    print(sol.findThePrefixCommonArray([2,3,1], [3,1,2]))


if __name__ == "__main__":
    main()

File: Leet 2657/solution.go
package main

import "fmt"

// Return the minimum of two integers
func minInt(num1 int, num2 int) int {
	if num1 < num2 {
		return num1
	}
	return num2
}

// Return the prefix common array between A and B (generic over comparable types)
func findThePrefixCommonArray[T comparable](A []T, B []T) []int {
	// each array contains distinct integers from 1 to n,
	// so if the same integer appears in both arrays,
	// it will be counted as a common integer

	length := minInt(len(A), len(B))
	result := make([]int, length)
	tally := make(map[T]int)

	var count int = 0
	for i := 0; i < length; i++ {
		tally[A[i]]++
		if tally[A[i]] == 2 {
			count++
		}

		tally[B[i]]++
		if tally[B[i]] == 2 {
			count++
		}

		result[i] = count
	}

	return result
}

func printArr[T any](message string, arr []T) {
	fmt.Printf("%s: [", message)
	for i := range arr {
		fmt.Printf(" %v ", arr[i])
	}
	fmt.Printf("]\n")
}

func main() {
	fmt.Println("2657. Find the Prefix Common Array of Two Arrays")

	arrayA := []int{1, 3, 2, 4}
	arrayB := []int{3, 1, 2, 4}

	var result []int = findThePrefixCommonArray(arrayA, arrayB)

	printArr("A", arrayA)
	printArr("A", arrayB)
	printArr("Result", result)
}
